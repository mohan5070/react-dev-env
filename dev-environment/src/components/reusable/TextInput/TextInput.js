/* eslint-disable jsx-a11y/label-has-for */
import React, { useState } from "react";
import PropTypes from "prop-types";
import Tooltip from "../Tooltip";
import styles from "./TextInput.module.scss";

/** Reusable Text Input with integrated label, error, and required field validation */
const TextInput = ({
  id,
  label,
  type,
  value,
  error,
  maxLength,
  required,
  onBlur,
  onChange
}) => {
  const [localError, setLocalError] = useState("");

  function handleBlur(event) {
    setLocalError(required && !value ? "Required." : "");
    if (onBlur) onBlur(event);
  }

  const hasError = error || localError;

  return (
    <>
      <label className={styles.label} htmlFor={id}>
        {label} {required && "*"}
      </label>
      <input
        type={type}
        value={value}
        required={required}
        maxLength={maxLength}
        className={styles.input + (hasError && " " + styles.error)}
        id={id}
        onBlur={handleBlur}
        onChange={onChange}
      />
      {localError && <Tooltip tip={localError} />}
    </>
  );
};

TextInput.propTypes = {
  /** Input ID */
  id: PropTypes.string.isRequired,

  /** Input label */
  label: PropTypes.string.isRequired,

  /** Input type */
  type: PropTypes.oneOf(["text", "email", "number"]).isRequired,

  /** Input value */
  value: PropTypes.string.isRequired,

  /** Function called onBlur */
  onBlur: PropTypes.func,

  /** Function called onChange */
  onChange: PropTypes.func.isRequired,

  /** Input max length */
  maxLength: PropTypes.number,

  /** Set to true to enable required field validation on blur */
  required: PropTypes.bool,

  /** Set error state and display below the input */
  error: PropTypes.string
};

TextInput.defaultProps = {
  required: false,
  error: "",
  type: "text"
};

export default TextInput;
