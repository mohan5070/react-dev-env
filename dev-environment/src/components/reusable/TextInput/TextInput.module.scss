/* Consider https://github.com/gajus/babel-plugin-react-css-modules */
@import "../../../variables";

.label {
  display: block;
  padding-bottom: 5px;
  font-size: $fontSizeMedium;
  color: $darkGray;
  font-family: $fontFamily;
  font-weight: bold;
}

.input {
  border: 1px solid $lightgray;
  padding: 0 10px;
  height: 44px;
  font-size: $fontSizeMedium;
  font-weight: 400;
}

.error {
  border-color: $red;
}
