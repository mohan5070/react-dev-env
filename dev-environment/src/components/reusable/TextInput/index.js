// Barrel for shortening imports
export { default } from "./TextInput";
