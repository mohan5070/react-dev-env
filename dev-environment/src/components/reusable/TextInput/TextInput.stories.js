import React, { useState } from "react";
import { storiesOf } from "@storybook/react";
import TextInput from "./TextInput";
import { text, boolean } from "@storybook/addon-knobs";
import { host } from "storybook-host";

// Consider https://github.com/Sambego/storybook-state instead
const TextInputExample = ({ value = "", required = false }) => {
  const [firstName, setFirstName] = useState(value);

  function handleChange(event) {
    setFirstName(event.target.value);
  }

  return (
    <TextInput
      label={text("label", "First Name")}
      value={text("value", firstName)}
      onChange={handleChange}
      required={boolean("required", required)}
      id="first-name"
    />
  );
};

storiesOf("TextInput", module)
  .addDecorator(
    host({
      title: "TextInput",
      align: "center bottom",
      height: "80%",
      width: 400
    })
  )
  .add("Docs (no change handler)", () => (
    <TextInput
      label={text("label", "First Name")}
      value={text("value", "")}
      onChange={() => {}}
      required={boolean("required", false)}
      id="first-name"
    />
  ))
  .add("Optional example", () => <TextInputExample />)
  .add("Required example", () => <TextInputExample required />);
