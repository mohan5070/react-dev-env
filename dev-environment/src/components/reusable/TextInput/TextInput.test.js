import React from "react";
import TextInput from "./TextInput";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer
    .create(<TextInput label="label" value="value" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
