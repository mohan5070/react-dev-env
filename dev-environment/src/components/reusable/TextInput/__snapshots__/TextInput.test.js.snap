// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <label
    className="label"
  >
    label
  </label>,
  <input
    className="input"
    onBlur={[Function]}
    required={false}
    value="value"
  />,
  "",
]
`;
